// pkg/lib-es5/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.exec = undefined;

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var needViaCache = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(target) {
    var s, c;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            s = stringifyTarget(target);
            c = targetsCache[s];

            if (!c) {
              _context.next = 4;
              break;
            }

            return _context.abrupt('return', c);

          case 4:
            _context.next = 6;
            return (0, _pkgFetch.need)(target);

          case 6:
            c = _context.sent;

            targetsCache[s] = c;
            return _context.abrupt('return', c);

          case 9:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function needViaCache(_x) {
    return _ref.apply(this, arguments);
  };
}();

var exec = exports.exec = function () {
  var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(argv2) {
    var argv, debug, forceBuild, input, inputJson, inputBin, bin, inputFin, config, configJson, output, outputDir, autoOutput, name, ext, targets, different, target, file, _iteratorNormalCompletion4, _didIteratorError4, _iteratorError4, _iterator4, _step4, _target, _file, options, _iteratorNormalCompletion5, _didIteratorError5, _iteratorError5, _iterator5, _step5, _target2, _iteratorNormalCompletion6, _didIteratorError6, _iteratorError6, _iterator6, _step6, _target3, f, tuple, records, stripes, _iteratorNormalCompletion7, _didIteratorError7, _iteratorError7, _iterator7, _step7, _target4, slash, _iteratorNormalCompletion8, _didIteratorError8, _iteratorError8, _iterator8, _step8, _target5;

    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            // eslint-disable-line complexity

            argv = (0, _minimist2.default)(argv2, {
              boolean: ['b', 'build', 'd', 'debug', 'h', 'help'],
              string: ['_', 'c', 'config', 'o', 'options', 'output', 'outdir', 'out-dir', 't', 'target', 'targets']
            });

            if (!(argv.h || argv.help)) {
              _context2.next = 4;
              break;
            }

            (0, _help2.default)();
            return _context2.abrupt('return');

          case 4:

            // debug

            debug = argv.d || argv.debug;

            _log.log.debugMode = debug;

            // forceBuild

            forceBuild = argv.b || argv.build;

            // _

            if (argv._.length) {
              _context2.next = 9;
              break;
            }

            throw (0, _log.wasReported)('Entry file/directory is expected', ['Pass --help to see usage information']);

          case 9:
            if (!(argv._.length > 1)) {
              _context2.next = 11;
              break;
            }

            throw (0, _log.wasReported)('Not more than one entry file/directory is expected');

          case 11:

            // input

            input = _path2.default.resolve(argv._[0]);
            _context2.next = 14;
            return (0, _fsPromise.exists)(input);

          case 14:
            if (_context2.sent) {
              _context2.next = 16;
              break;
            }

            throw (0, _log.wasReported)('Input file ' + input + ' does not exist');

          case 16:
            _context2.next = 18;
            return (0, _fsPromise.stat)(input);

          case 18:
            if (!_context2.sent.isDirectory()) {
              _context2.next = 24;
              break;
            }

            input = _path2.default.join(input, 'package.json');
            _context2.next = 22;
            return (0, _fsPromise.exists)(input);

          case 22:
            if (_context2.sent) {
              _context2.next = 24;
              break;
            }

            throw (0, _log.wasReported)('Input file ' + input + ' does not exist');

          case 24:

            // inputJson

            inputJson = null;

            if (!(0, _common.isPackageJson)(input)) {
              _context2.next = 31;
              break;
            }

            _context2.t0 = JSON;
            _context2.next = 29;
            return (0, _fsPromise.readFile)(input);

          case 29:
            _context2.t1 = _context2.sent;
            inputJson = _context2.t0.parse.call(_context2.t0, _context2.t1);

          case 31:

            // inputBin

            inputBin = null;

            if (!inputJson) {
              _context2.next = 41;
              break;
            }

            bin = inputJson.bin;

            if (!bin) {
              _context2.next = 41;
              break;
            }

            if ((typeof bin === 'undefined' ? 'undefined' : (0, _typeof3.default)(bin)) === 'object') {
              if (bin[inputJson.name]) {
                bin = bin[inputJson.name];
              } else {
                bin = bin[(0, _keys2.default)(bin)[0]]; // TODO multiple inputs to compile them all?
              }
            }
            inputBin = _path2.default.resolve(_path2.default.dirname(input), bin);
            _context2.next = 39;
            return (0, _fsPromise.exists)(inputBin);

          case 39:
            if (_context2.sent) {
              _context2.next = 41;
              break;
            }

            throw (0, _log.wasReported)('Package.json \'bin\' property ' + inputBin + ' does not exist');

          case 41:

            // inputFin

            inputFin = inputBin || input;

            // config

            config = argv.c || argv.config;

            if (!(inputJson && config)) {
              _context2.next = 45;
              break;
            }

            throw (0, _log.wasReported)('Specify either \'package.json\' or config. Not both');

          case 45:

            // configJson

            configJson = null;

            if (!config) {
              _context2.next = 54;
              break;
            }

            config = _path2.default.resolve(config);
            _context2.next = 50;
            return (0, _fsPromise.exists)(config);

          case 50:
            if (_context2.sent) {
              _context2.next = 52;
              break;
            }

            throw (0, _log.wasReported)('Config file ' + config + ' does not exist');

          case 52:
            configJson = require(config); // may be either json or js
            if (!configJson.name && !configJson.files && !configJson.dependencies && !configJson.pkg) {
              // package.json not detected
              configJson = { pkg: configJson };
            }

          case 54:

            // output, outputDir

            output = argv.o || argv.output;
            outputDir = argv.outdir || argv['out-dir'];
            autoOutput = false;

            if (!(output && outputDir)) {
              _context2.next = 59;
              break;
            }

            throw (0, _log.wasReported)('Specify either \'output\' or \'out-dir\'. Not both');

          case 59:
            if (output) {
              _context2.next = 72;
              break;
            }

            name = void 0;

            if (!inputJson) {
              _context2.next = 67;
              break;
            }

            name = inputJson.name;

            if (name) {
              _context2.next = 65;
              break;
            }

            throw (0, _log.wasReported)('Property \'name\' does not exist in ' + argv._[0]);

          case 65:
            _context2.next = 68;
            break;

          case 67:
            name = _path2.default.basename(inputFin);

          case 68:
            autoOutput = true;
            ext = _path2.default.extname(name);

            output = name.slice(0, -ext.length || undefined);
            output = _path2.default.resolve(outputDir || '', output);

          case 72:

            // targets

            targets = parseTargets(argv.t || argv.target || argv.targets);


            if (!targets.length) {
              if (!autoOutput) {
                targets = parseTargets('host');
                (0, _assert2.default)(targets.length === 1);
              } else {
                targets = parseTargets('linux,macos,win');
              }
              _log.log.info('Targets not specified. Assuming:', '' + targets.map(stringifyTarget).join(', '));
            }

            // differentParts

            different = differentParts(targets);

            // targets[].output

            if (!(targets.length === 1)) {
              _context2.next = 82;
              break;
            }

            target = targets[0];
            file = output;

            if (target.platform === 'win' && autoOutput) file += '.exe';
            target.output = file;
            _context2.next = 101;
            break;

          case 82:
            _iteratorNormalCompletion4 = true;
            _didIteratorError4 = false;
            _iteratorError4 = undefined;
            _context2.prev = 85;

            for (_iterator4 = (0, _getIterator3.default)(targets); !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
              _target = _step4.value;
              _file = stringifyTargetForOutput(output, _target, different);

              if (_target.platform === 'win') _file += '.exe';
              _target.output = _file;
            }
            _context2.next = 93;
            break;

          case 89:
            _context2.prev = 89;
            _context2.t2 = _context2['catch'](85);
            _didIteratorError4 = true;
            _iteratorError4 = _context2.t2;

          case 93:
            _context2.prev = 93;
            _context2.prev = 94;

            if (!_iteratorNormalCompletion4 && _iterator4.return) {
              _iterator4.return();
            }

          case 96:
            _context2.prev = 96;

            if (!_didIteratorError4) {
              _context2.next = 99;
              break;
            }

            throw _iteratorError4;

          case 99:
            return _context2.finish(96);

          case 100:
            return _context2.finish(93);

          case 101:

            // options

            options = argv.options || '';

            options = options.split(',').filter(function (option) {
              return option;
            }).map(function (option) {
              return '--' + option;
            });

            // check if input is going
            // to be overwritten by output

            _iteratorNormalCompletion5 = true;
            _didIteratorError5 = false;
            _iteratorError5 = undefined;
            _context2.prev = 106;
            _iterator5 = (0, _getIterator3.default)(targets);

          case 108:
            if (_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done) {
              _context2.next = 119;
              break;
            }

            _target2 = _step5.value;

            if (!(_target2.output === inputFin)) {
              _context2.next = 116;
              break;
            }

            if (!autoOutput) {
              _context2.next = 115;
              break;
            }

            _target2.output += '-' + _target2.platform;
            _context2.next = 116;
            break;

          case 115:
            throw (0, _log.wasReported)('Refusing to overwrite input ' + inputFin);

          case 116:
            _iteratorNormalCompletion5 = true;
            _context2.next = 108;
            break;

          case 119:
            _context2.next = 125;
            break;

          case 121:
            _context2.prev = 121;
            _context2.t3 = _context2['catch'](106);
            _didIteratorError5 = true;
            _iteratorError5 = _context2.t3;

          case 125:
            _context2.prev = 125;
            _context2.prev = 126;

            if (!_iteratorNormalCompletion5 && _iterator5.return) {
              _iterator5.return();
            }

          case 128:
            _context2.prev = 128;

            if (!_didIteratorError5) {
              _context2.next = 131;
              break;
            }

            throw _iteratorError5;

          case 131:
            return _context2.finish(128);

          case 132:
            return _context2.finish(125);

          case 133:

            // fetch targets

            _iteratorNormalCompletion6 = true;
            _didIteratorError6 = false;
            _iteratorError6 = undefined;
            _context2.prev = 136;
            _iterator6 = (0, _getIterator3.default)(targets);

          case 138:
            if (_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done) {
              _context2.next = 155;
              break;
            }

            _target3 = _step6.value;

            _target3.forceBuild = forceBuild;
            _context2.next = 143;
            return needViaCache(_target3);

          case 143:
            _target3.binaryPath = _context2.sent;
            f = _target3.fabricator = fabricatorForTarget(_target3);

            f.forceBuild = forceBuild;
            _context2.next = 148;
            return needViaCache(f);

          case 148:
            f.binaryPath = _context2.sent;

            if (!(f.platform !== 'win')) {
              _context2.next = 152;
              break;
            }

            _context2.next = 152;
            return (0, _chmod.plusx)(f.binaryPath);

          case 152:
            _iteratorNormalCompletion6 = true;
            _context2.next = 138;
            break;

          case 155:
            _context2.next = 161;
            break;

          case 157:
            _context2.prev = 157;
            _context2.t4 = _context2['catch'](136);
            _didIteratorError6 = true;
            _iteratorError6 = _context2.t4;

          case 161:
            _context2.prev = 161;
            _context2.prev = 162;

            if (!_iteratorNormalCompletion6 && _iterator6.return) {
              _iterator6.return();
            }

          case 164:
            _context2.prev = 164;

            if (!_didIteratorError6) {
              _context2.next = 167;
              break;
            }

            throw _iteratorError6;

          case 167:
            return _context2.finish(164);

          case 168:
            return _context2.finish(161);

          case 169:

            // tuple

            tuple = void 0;


            if (configJson) {
              tuple = {
                config: configJson,
                base: _path2.default.dirname(config)
              };
            } else {
              tuple = {
                config: inputJson || {},
                base: _path2.default.dirname(input) // not `inputBin` because only `input`
              }; // is the place for `inputJson`
            }

            // records

            _context2.next = 173;
            return (0, _walker2.default)({
              tuple: tuple, input: inputFin
            });

          case 173:
            records = _context2.sent;
            stripes = {};
            _iteratorNormalCompletion7 = true;
            _didIteratorError7 = false;
            _iteratorError7 = undefined;
            _context2.prev = 178;
            _iterator7 = (0, _getIterator3.default)(targets);

          case 180:
            if (_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done) {
              _context2.next = 191;
              break;
            }

            _target4 = _step7.value;
            slash = _target4.platform === 'win' ? '\\' : '/';

            _target4.slash = slash;

            if (stripes[slash]) {
              _context2.next = 188;
              break;
            }

            _context2.next = 187;
            return (0, _packer2.default)({ records: records, slash: slash });

          case 187:
            stripes[slash] = _context2.sent;

          case 188:
            _iteratorNormalCompletion7 = true;
            _context2.next = 180;
            break;

          case 191:
            _context2.next = 197;
            break;

          case 193:
            _context2.prev = 193;
            _context2.t5 = _context2['catch'](178);
            _didIteratorError7 = true;
            _iteratorError7 = _context2.t5;

          case 197:
            _context2.prev = 197;
            _context2.prev = 198;

            if (!_iteratorNormalCompletion7 && _iterator7.return) {
              _iterator7.return();
            }

          case 200:
            _context2.prev = 200;

            if (!_didIteratorError7) {
              _context2.next = 203;
              break;
            }

            throw _iteratorError7;

          case 203:
            return _context2.finish(200);

          case 204:
            return _context2.finish(197);

          case 205:

            _log.log.debug('Targets:', (0, _stringify2.default)(targets));

            _iteratorNormalCompletion8 = true;
            _didIteratorError8 = false;
            _iteratorError8 = undefined;
            _context2.prev = 209;
            _iterator8 = (0, _getIterator3.default)(targets);

          case 211:
            if (_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done) {
              _context2.next = 225;
              break;
            }

            _target5 = _step8.value;
            _context2.next = 215;
            return (0, _fsPromise.mkdirp)(_path2.default.dirname(_target5.output));

          case 215:
            _context2.next = 217;
            return (0, _fsPromise.remove)(_target5.output);

          case 217:
            _context2.next = 219;
            return (0, _producer2.default)({ stripe: stripes[_target5.slash], options: options, target: _target5 });

          case 219:
            if (!(_target5.platform !== 'win')) {
              _context2.next = 222;
              break;
            }

            _context2.next = 222;
            return (0, _chmod.plusx)(_target5.output);

          case 222:
            _iteratorNormalCompletion8 = true;
            _context2.next = 211;
            break;

          case 225:
            _context2.next = 231;
            break;

          case 227:
            _context2.prev = 227;
            _context2.t6 = _context2['catch'](209);
            _didIteratorError8 = true;
            _iteratorError8 = _context2.t6;

          case 231:
            _context2.prev = 231;
            _context2.prev = 232;

            if (!_iteratorNormalCompletion8 && _iterator8.return) {
              _iterator8.return();
            }

          case 234:
            _context2.prev = 234;

            if (!_didIteratorError8) {
              _context2.next = 237;
              break;
            }

            throw _iteratorError8;

          case 237:
            return _context2.finish(234);

          case 238:
            return _context2.finish(231);

          case 239:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this, [[85, 89, 93, 101], [94,, 96, 100], [106, 121, 125, 133], [126,, 128, 132], [136, 157, 161, 169], [162,, 164, 168], [178, 193, 197, 205], [198,, 200, 204], [209, 227, 231, 239], [232,, 234, 238]]);
  }));

  return function exec(_x2) {
    return _ref2.apply(this, arguments);
  };
}();

var _fsPromise = require('fs-promise');

var _log = require('./log.js');

var _pkgFetch = require('pkg-fetch');

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _help = require('./help');

var _help2 = _interopRequireDefault(_help);

var _common = require('../prelude/common.js');

var _minimist = require('minimist');

var _minimist2 = _interopRequireDefault(_minimist);

var _packer = require('./packer.js');

var _packer2 = _interopRequireDefault(_packer);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _chmod = require('./chmod.js');

var _producer = require('./producer.js');

var _producer2 = _interopRequireDefault(_producer);

var _walker = require('./walker.js');

var _walker2 = _interopRequireDefault(_walker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// http://www.openwall.com/lists/musl/2012/12/08/4

var hostArch = _pkgFetch.system.hostArch,
    hostPlatform = _pkgFetch.system.hostPlatform,
    knownArchs = _pkgFetch.system.knownArchs,
    knownPlatforms = _pkgFetch.system.knownPlatforms,
    toFancyArch = _pkgFetch.system.toFancyArch,
    toFancyPlatform = _pkgFetch.system.toFancyPlatform;

var hostNodeRange = 'node' + process.version.match(/^v(\d+)/)[1];

function parseTargets(items) {
  // node6-macos-x64,node6-linux-x64
  if (!items) return [];
  if (Array.isArray(items)) {
    throw (0, _log.wasReported)('Something is wrong near ' + items[0]);
  }
  var targets = [];
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = (0, _getIterator3.default)(items.split(',')), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var item = _step.value;

      var target = {
        nodeRange: hostNodeRange,
        platform: hostPlatform,
        arch: hostArch
      };
      if (item !== 'host') {
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          for (var _iterator2 = (0, _getIterator3.default)(item.split('-')), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var token = _step2.value;

            if (!token) continue;
            var p = toFancyPlatform(token);
            if (knownPlatforms.indexOf(p) >= 0) {
              target.platform = p;
              continue;
            }
            var a = toFancyArch(token);
            if (knownArchs.indexOf(a) >= 0) {
              target.arch = a;
              continue;
            }
            target.nodeRange = token;
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }
      }
      targets.push(target);
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return targets;
}

function stringifyTarget(target) {
  var nodeRange = target.nodeRange,
      platform = target.platform,
      arch = target.arch;

  return nodeRange + '-' + platform + '-' + arch;
}

function differentParts(targets) {
  var nodeRanges = {};
  var platforms = {};
  var archs = {};
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = (0, _getIterator3.default)(targets), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var target = _step3.value;

      nodeRanges[target.nodeRange] = true;
      platforms[target.platform] = true;
      archs[target.arch] = true;
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3.return) {
        _iterator3.return();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  var result = {};
  if ((0, _keys2.default)(nodeRanges).length > 1) {
    result.nodeRange = true;
  }
  if ((0, _keys2.default)(platforms).length > 1) {
    result.platform = true;
  }
  if ((0, _keys2.default)(archs).length > 1) {
    result.arch = true;
  }
  return result;
}

function stringifyTargetForOutput(output, target, different) {
  var a = [output];
  if (different.nodeRange) a.push(target.nodeRange);
  if (different.platform) a.push(target.platform);
  if (different.arch) a.push(target.arch);
  return a.join('-');
}

function fabricatorForTarget(target) {
  var nodeRange = target.nodeRange,
      arch = target.arch;

  return { nodeRange: nodeRange, platform: hostPlatform, arch: arch };
}

var targetsCache = {};
